lib/grp.gi

```diff
@@ -483,27 +484,35 @@ local p, hom, reps, a, b, ap_bp, ab_p, H;
   reps := Filtered(reps, g -> not IsOne(g));
   reps := List(reps, g -> PreImagesRepresentative(hom, g));
 
-  for b in Image(hom) do
-    b := PreImagesRepresentative(hom, b);
-    for a in reps do
+  as := List(reps, a -> [a,a^p]);;
+  bs := List(Image(hom), function(b)
+      b := PreImagesRepresentative(hom, b);
+      return [b,b^p];
+      end);;
+
+  for b in bs do
+    bp := b[2]; b := b[1];
+    for a in as do
+      ap := a[2]; a := a[1];
       # if a and b commute the regularity condition automatically holds
-      if a*b = b*a then continue; fi;
+      ab := a*b;
+      if ab = b*a then continue; fi;
 
       # regularity is also automatic if a^p * b^p = (a*b)^p
-      ap_bp := a^p * b^p;
-      ab_p := (a*b)^p;
+      ap_bp := ap * bp;
+      ab_p := ab^p;
       if ap_bp = ab_p then continue; fi;
 
       # if the subgroup generated H by a and b is itself regular, we are also
-      # done. However we don't use recursion, here, as H may be equal to G;
-      # and also we have to be careful to not use too expensive code here.
-      # But a quick size check is certainly fine.
+      # done. However we don't use recursion here, as it is too expensive.
+      # we just check the direct definition, with a minor twist to avoid
+      # calling Agemo
+      g := ap_bp / ab_p;
+      h := Comm(a,b)^p;
+      if g = h or IsOne(g*h) then continue; fi;
```

Comment:
Good. This new check is most advantageous for small p, but should always be cheap (g=h always cheap, g*h should not be expensive). For small p, I suspect it completely avoids the subgroup check in many cases

Reply:
Thank you very, very much for you feedback @jackschmidt and sorry for taking so long to get back to you.

Regarding this point, indeed `g=h` is fast and indeed for small $p$ the cost for checking `IsOne(g*h)` is absolutely worth it, based on extensive tests with groups of order $3^{10}$ formed as direct products of smaller $3$-groups.